### ISSUE TITLE
No delay on dictionary close (after word is added to vocabulary builder)

### ISSUE THREAD

**Issue by sdothum** [koreader/koreader#12250]
* KOReader version: 2024.07
* Device: Kobo Libra Colour

#### Issue

On a dictionary lookup, if the word is "added to vocabulary builder", on close of the dictionary widget, the highlighted selection is cleared immediately (no 0.5sec delay). Simple dictionary lookup actions work as expected with the 0.5sec delay.

#### Steps to reproduce

- long press word to highlight and open dictionary
- press "add to vocabulary builder"
- close dictionary widget

**Comment by sdothum** [comment]
This appears to have "self-healed" by simply toggling any setting in Screen -> E-ink settings -> Full refresh rate -> (previous settings..).

**Comment by sdothum** [comment]
Spoke too soon.. the issue returns after awhile.. even during the time reading on the same page.

My only guess is that something changed in the e-ink refresh logic?

**Comment by NiLuJe** [comment]
Someone might be using `partial` instead of `ui`/`flashui` and you get to see a promotion to flash on something that should really never have been promoted.

Unclear, as I don't use any of the features mentioned, so have no idea how they ought to behave.

Keep in mind, this is a color device, and color brings with it its own wonderful set of weird refresh quirks.

(On that note, do try a recent nightly, as color-specific codepaths have been updated a few days ago).

**Comment by NiLuJe** [comment]
As usual: please don't ignore the issue template and provide verbose debug logs, that would have answered my question already ;).

**Comment by jonnyl2** [comment]
It seems to be related to this scheduled 0.5s delay in clearing the selection upon closing the dictionary dialog, which for some reason does not get called after pressing the 'Add to Vocab builder' button: 

https://github.com/koreader/koreader/blob/f9ee2449c7923df75b75961a50cea64990585fa3/frontend/ui/widget/dictquicklookup.lua#L1145-L1152

**Comment by sdothum** [comment]
[crash.log](https://github.com/user-attachments/files/16457627/crash.log)

Attached crash log (startup trimmed off).

Crash log contains 2 dictionary lookups (text = "turrets"):

First is simple dictionary lookup and widget close.

Second is dictionary lookup, "add to vocabulary builder",  and widget close.

As previously described, the simple (first) lookup executes DictQuickLookup.onClose function's delay successfully (though, i'm not conversant enough to identify it reference in the crash.log)

Not sure what happens when the vocabulary bulider is accessed -- why it generally causes the dictionary widget to close the highlighted word immediately as the widget itself is closed, and why, if one toggles any of the "Full Refresh Rate" settings, the delay works momentarily.

Hope this helps. (FYI prior to 2024.07, the earlier nightly patched 2024.04 for the Kobo Libra Colour, had no issues with the highlight delay after adding a word to the vocabulary builder).

**Comment by NiLuJe** [comment]
At a quick glance, the refreshes look sound, so I'm going to assume @jonnyl2 is on the right track, and flag this as plugin-related (e.g., vocab) somehow...

The fact that it didn't behave that way in 2024.07 is potentially good to know, but I'm not familiar enough with the code in question to tell exactly why...

**Comment by poire-z** [comment]
No, nothing related to scheduleIn, which gets triggered 0.5s later - but the highlight has already been removed.
It's:
https://github.com/koreader/koreader/blob/7ef78e606e168b494c31b7a11714bffb2b6c6274/plugins/vocabbuilder.koplugin/main.lua#L2060-L2067
that calls:
https://github.com/koreader/koreader/blob/7ef78e606e168b494c31b7a11714bffb2b6c6274/frontend/document/credocument.lua#L769-L789
and it's actually the getTextFromXPointers() on lines 785 & 786 that, even if asked to not draw the segment from start to end, must have crengine clear any past "selection".
This may be some internal normal behaviour, as it is only made to ever handle/draw only one selection/segment - when we call these again to get context, it must lose the previous lookupword selection and not have it drawn when refresh/repaint the page.

I don't think I'll ever want to fix that :/ it feels like it would need quite some kludge needed for this edge case.

**Comment by poire-z** [comment]
Or that bit of code could, after clearing and getting the context bits, re-select the original word :)
```diff
--- a/frontend/document/credocument.lua
+++ b/frontend/document/credocument.lua
@@ -785,6 +785,11 @@ function CreDocument:getSelectedWordContext(word, nb_words, pos0, pos1)
     local prev = self:getTextFromXPointers(pos_start, pos0)
     local next = self:getTextFromXPointers(pos1, pos_end)

+    -- getTextFromXPointers() may have cleared the original selection,
+    -- and crengine would then not draw it on page refresh.
+    -- So, have it know again about what should be selected and drawn.
+    self:getTextFromXPointers(pos0, pos1, true)
+
     return prev, next
 end

```

**Comment by sdothum** [comment]
This patch appears to be a livable solution (in lieu of my being able to assess the changes to the vocabulary builder plugin from 2024.04 to 2024.07). 

There is a side-effect:

the "Add to vocabulary builder" button on the dictionary widget does not flash properly (the button does not invert in color uniformly -- just a partial incomplete refresh -- but it does dim as expected after being triggered). i can live with this as it is better than losing one's page position when the highlight is cleared under the dictionary widget as it is being closed :-)

Will give this a good shakedown today and keep you posted.

THANK YOU!

**Comment by poire-z** [comment]
> Hope this helps. (FYI prior to 2024.07, the earlier nightly patched 2024.04 for the Kobo Libra Colour, had no issues with the highlight delay after adding a word to the vocabulary builder).

You're sure about that?
The involved code seems to be from 2022.
Haven't you enabled between 2024.04 and 2024.07 something related to saving word context?
Otherwise, the only big PR merged in between is #11777.

**Comment by sdothum** [comment]
> You're sure about that? The involved code seems to be from 2022. Haven't you enabled between 2024.04 and 2024.07 something related to saving word context? Otherwise, the only big PR merged in between is #11777.

My septuagenarian memory would have sworn so... but i went back and re-installed 2024.04 and applied the patches from #11737 just to be sure..

..i stand corrected. With that release and Kobo Libra Colour #11737 patches, the "Add to vocabulary builder" also exhibited the immediate clear of the highlighted dictionary word upon closing the dictionary widget.

FYI i have made an effort since yesterday to perform an excessive number of dictionary lookups with vocabulary builder "adds" switching amoungst several books and your patch above works perfectly.

**Comment by Commodore64user** [comment]
> Or that bit of code could, after clearing and getting the context bits, re-select the original word :)
> 
> ```diff
> 
> --- a/frontend/document/credocument.lua
> 
> +++ b/frontend/document/credocument.lua
> 
> @@ -785,6 +785,11 @@ function CreDocument:getSelectedWordContext(word, nb_words, pos0, pos1)
> 
>      local prev = self:getTextFromXPointers(pos_start, pos0)
> 
>      local next = self:getTextFromXPointers(pos1, pos_end)
> 
> 
> 
> +    -- getTextFromXPointers() may have cleared the original selection,
> 
> +    -- and crengine would then not draw it on page refresh.
> 
> +    -- So, have it know again about what should be selected and drawn.
> 
> +    self:getTextFromXPointers(pos0, pos1, true)
> 
> +
> 
>      return prev, next
> 
>  end
> 
> 
> 
> ```
> 
> 

If only we could do the same re-select for words after a full search is performed.
